Add Gloo OpenSSL-over-TCP transport

pbelevich opened the pull request:
This PR adds [Gloo OpenSSL-over-TCP transport](https://github.com/facebookincubator/gloo/pull/260) to PyTorch. 

- [x] Update `third_party/gloo` to [OpenSSL-over-TCP transport](01e2c2660cd43963ce1fe3e21220ac01f07d9a4b)
- [x] `find_package(OpenSSL REQUIRED)` and building Gloo with `USE_TCP_OPENSSL`
- [x] Update `torch/lib/c10d/GlooDeviceFactory.cpp` with new `TCP_TLS` key that will create Gloo TLS transport
- [ ] Make OpenSSL(libssl-dev) available on all CI linux environments, so CMake would be able to find it
- [ ] Add new CI test(not build) workflow that 
  - Gets fresh pytorch build
  - Generates `pkey.key`, `cert.pem` and `ca.pem`:

```
openssl genrsa -out ca.key 2048
openssl req -subj "/C=US/ST=New York/L=New York/O=Gloo Certificate Authority" -new -x509 -days 7300 -key ca.key -sha256 -extensions v3_ca -out ca.pem

openssl genrsa -out pkey.key 2048
openssl req -subj "/C=US/ST=California/L=San Francisco/O=Gloo Testing Company" -sha256 -new -key pkey.key -out csr.csr

openssl x509 -sha256 -req -in csr.csr -CA ca.pem -CAkey ca.key -CAcreateserial -out cert.pem -days 7300

openssl verify -CAfile ca.pem cert.pem
```
  - Runs `./build/bin/ProcessGroupGlooTest`, `./build/bin/ProcessGroupGlooAsyncTest` and `pytorch/test/distributed/test*.py` with the following environment variables:

```
GLOO_DEVICE_TRANSPORT=TCP_TLS \
GLOO_DEVICE_TRANSPORT_TCP_TLS_PKEY=~/pkey.key \
GLOO_DEVICE_TRANSPORT_TCP_TLS_CERT=~/cert.pem \
GLOO_DEVICE_TRANSPORT_TCP_TLS_CA_FILE=~/ca.pem \
```

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#43482 Add Gloo OpenSSL-over-TCP transport**
* #50787 Install libssl-dev on Linux CI for Gloo TCP-TLS transport

pbelevich committed:
Add Gloo OpenSSL-over-TCP transport

[ghstack-poisoned]

mrshenli reviewed (commented):
Thanks for adding this. 

> This PR adds Gloo OpenSSL-over-TCP transport to PyTorch. This functionality works only on Linux and gh tool and openssl tool are required for manual testing.

How large are those key/cert/etc.? If they are not that large, can we include them in the tests so that we can have CI to ensure that future changes do not break it?

mrshenli commented on torch/lib/c10d/GlooDeviceFactory.cpp:
```diff
@@ -8,6 +8,10 @@
 #include <gloo/transport/tcp/device.h>
 #endif
 
+#if GLOO_HAVE_TRANSPORT_TCP_TLS
```
where is this defined?

pbelevich replied:
Defined in [gloo/config.h.in:37](https://github.com/facebookincubator/gloo/pull/260/files#diff-36d25b293ab5437090249cefcea89a2bR37) and set in [gloo/CMakeLists.txt:81-85](https://github.com/facebookincubator/gloo/pull/260/files#diff-3ec7c720473f942e710c60b0ea8d794fR82)

mrshenli commented on torch/lib/c10d/GlooDeviceFactory.cpp:
```diff
@@ -59,6 +63,31 @@ C10_REGISTER_CREATOR(GlooDeviceRegistry, LINUX, makeTCPDevice);
+  return ::gloo::transport::tcp::tls::CreateDevice(attr, pkey, cert, caFile, caPath);
+}
+
+C10_REGISTER_CREATOR(GlooDeviceRegistry, TCP_TLS, makeTCPTLSDevice);
```
where is `TCP_TLS` introduced?

pbelevich replied:
In thin particular line I believe

mrshenli reviewed (commented):
Which PR introduced `USE_TCP_OPENSSL` flag?

pbelevich commented:
> Which PR introduced `USE_TCP_OPENSSL` flag?
[facebookincubator/gloo#260](https://github.com/facebookincubator/gloo/pull/260/files#diff-af3b638bc2a3e6c650974192a53c7291R37)

pbelevich commented:
> How large are those key/cert/etc.? If they are not that large, can we include them in the tests so that we can have CI to ensure that future changes do not break it?
Less than 2kb each, I will think about it

pbelevich commented:
@mrshenli @malfet What is the best way to setup circleci to test this functionality? build all linux workflows with `USE_TCP_OPENSSL=ON` and run all c10d tests with `GLOO_DEVICE_TRANSPORT=TCP_TLS` and without it(the same as `GLOO_DEVICE_TRANSPORT=TCP`) or setup new separate workflow from  scratch?

mrshenli commented:
> build all linux workflows with USE_TCP_OPENSSL=ON and run all c10d tests with GLOO_DEVICE_TRANSPORT=TCP_TLS and without it(the same as GLOO_DEVICE_TRANSPORT=TCP)

This might slowdown existing CI quite a bit. @malfet does it make sense to create a dedicated workflow?

@pbelevich question, when compiled with `USE_TCP_OPENSSL=ON`, is the non-ssl TCP still available as an option or is it overwritten by TCP_TLS?

pbelevich commented:
> @pbelevich question, when compiled with `USE_TCP_OPENSSL=ON`, is the non-ssl TCP still available as an option or is it overwritten by TCP_TLS?

@mrshenli TCP is still available with `USE_TCP_OPENSSL=ON`

pbelevich commented:
> This might slowdown existing CI quite a bit. @malfet does it make sense to create a dedicated workflow?

IMHO, it will laterally add an extra second: compile 3 additional cpp files and run 1 or 2 additional tests

malfet commented:
> > This might slowdown existing CI quite a bit. @malfet does it make sense to create a dedicated workflow?
> 
> IMHO, it will laterally add an extra second: compile 3 additional cpp files and run 1 or 2 additional tests
In that case, I don't think it warrants a separate config, but I would recommend enabling it for just 1 or 2 of the available configs (to avoid regressions when code can no longer be compiled on Linux if that option is not set)

mrshenli commented:
> IMHO, it will laterally add an extra second: compile 3 additional cpp files and run 1 or 2 additional tests

I see. So you don't plan to run all existing c10d/DDP/RPC tests on the new backend?

pbelevich commented:
> I see. So you don't plan to run all existing c10d/DDP/RPC tests on the new backend?

@mrshenli actually just duplicate [this one line](https://github.com/pytorch/pytorch/blob/eaace3e10e1e6fef779d7045ef4f9093752e2139/.jenkins/pytorch/test.sh#L239) with `GLOO_DEVICE_TRANSPORT=TCP_TLS` ...

mrshenli commented:
> I see. So you don't plan to run all existing c10d/DDP/RPC tests on the new backend?

That is for C++-only tests. We have many more python tests. See `test_c10d.py`, `test_c10d_spawn.py`, `test_faulty_agent`.py, `test_process_group_agent.py`, `test_distributed.py`, etc. Python tests are included in https://github.com/pytorch/pytorch/blob/e08e93f94647fa2cfe46c62154d926d0b3af223b/test/run_test.py

mrshenli commented:
I assume users would use this feature from Python applications. If that's the case, we probably should also add some Python tests? Basically test in the way of how users use it.

pbelevich committed:
Update on "Add Gloo OpenSSL-over-TCP transport"

This PR adds [Gloo OpenSSL-over-TCP transport](https://github.com/facebookincubator/gloo/pull/260) to PyTorch. This functionality works only on Linux and [gh tool](https://github.com/cli/cli#installation) and [openssl tool](https://www.openssl.org/) are required for manual testing.

0. Generate `pkey.key`, `cert.pem` and `ca.pem` in the `$HOME` dir:
```
openssl genrsa -out ca.key 2048
openssl req -subj "/C=US/ST=New York/L=New York/O=Gloo Certificate Authority" -new -x509 -days 7300 -key ca.key -sha256 -extensions v3_ca -out ca.pem

openssl genrsa -out pkey.key 2048
openssl req -subj "/C=US/ST=California/L=San Francisco/O=Gloo Testing Company" -sha256 -new -key pkey.key -out csr.csr

openssl x509 -sha256 -req -in csr.csr -CA ca.pem -CAkey ca.key -CAcreateserial -out cert.pem -days 7300

openssl verify -CAfile ca.pem cert.pem
```
1. Checkout this PR (`gh pr checkout 43482`)
2. (skip if https://github.com/facebookincubator/gloo/pull/260 is already merged) Goto gloo subdir: `cd third_party/gloo`
3. (skip if https://github.com/facebookincubator/gloo/pull/260 is already merged) Checkout gloo's PR: `gh pr checkout 260` and go back `cd ../..`
4. Build PyTorch with `USE_TCP_OPENSSL=ON python setup.py develop`
5. Run ProcessGroupGlooTest:
```
GLOO_DEVICE_TRANSPORT=TCP_TLS \
GLOO_DEVICE_TRANSPORT_TCP_TLS_PKEY=~/pkey.key \
GLOO_DEVICE_TRANSPORT_TCP_TLS_CERT=~/cert.pem \
GLOO_DEVICE_TRANSPORT_TCP_TLS_CA_FILE=~/ca.pem \
./build/bin/ProcessGroupGlooTest
```
6. Run ProcessGroupGlooAsyncTest:
```
GLOO_DEVICE_TRANSPORT=TCP_TLS \
GLOO_DEVICE_TRANSPORT_TCP_TLS_PKEY=~/pkey.key \
GLOO_DEVICE_TRANSPORT_TCP_TLS_CERT=~/cert.pem \
GLOO_DEVICE_TRANSPORT_TCP_TLS_CA_FILE=~/ca.pem \
./build/bin/ProcessGroupGlooAsyncTest
```




[ghstack-poisoned]

mrshenli reviewed (commented):
Follow up with today's discussion. Other options to enable this feature including:

1. allow users to specify `backend="gloo_tls"` through the [`init_process_group`](https://pytorch.org/docs/stable/distributed.html#torch.distributed.init_process_group) API, and read env vars to get paths to key/cert files.
2. use `backend="gloo_tls"` and` init_method="tcp://xxxxx:yy?key=aaaa&cert=bbbb"`. 

There might be other options.

pbelevich committed:
Update on "Add Gloo OpenSSL-over-TCP transport"

This PR adds [Gloo OpenSSL-over-TCP transport](https://github.com/facebookincubator/gloo/pull/260) to PyTorch. This functionality works only on Linux and [gh tool](https://github.com/cli/cli#installation) and [openssl tool](https://www.openssl.org/) are required for manual testing.

0. Generate `pkey.key`, `cert.pem` and `ca.pem` in the `$HOME` dir:
```
openssl genrsa -out ca.key 2048
openssl req -subj "/C=US/ST=New York/L=New York/O=Gloo Certificate Authority" -new -x509 -days 7300 -key ca.key -sha256 -extensions v3_ca -out ca.pem

openssl genrsa -out pkey.key 2048
openssl req -subj "/C=US/ST=California/L=San Francisco/O=Gloo Testing Company" -sha256 -new -key pkey.key -out csr.csr

openssl x509 -sha256 -req -in csr.csr -CA ca.pem -CAkey ca.key -CAcreateserial -out cert.pem -days 7300

openssl verify -CAfile ca.pem cert.pem
```
1. Checkout this PR (`gh pr checkout 43482`)
2. (skip if https://github.com/facebookincubator/gloo/pull/260 is already merged) Goto gloo subdir: `cd third_party/gloo`
3. (skip if https://github.com/facebookincubator/gloo/pull/260 is already merged) Checkout gloo's PR: `gh pr checkout 260` and go back `cd ../..`
4. Build PyTorch with `USE_TCP_OPENSSL=ON python setup.py develop`
5. Run ProcessGroupGlooTest:
```
GLOO_DEVICE_TRANSPORT=TCP_TLS \
GLOO_DEVICE_TRANSPORT_TCP_TLS_PKEY=~/pkey.key \
GLOO_DEVICE_TRANSPORT_TCP_TLS_CERT=~/cert.pem \
GLOO_DEVICE_TRANSPORT_TCP_TLS_CA_FILE=~/ca.pem \
./build/bin/ProcessGroupGlooTest
```
6. Run ProcessGroupGlooAsyncTest:
```
GLOO_DEVICE_TRANSPORT=TCP_TLS \
GLOO_DEVICE_TRANSPORT_TCP_TLS_PKEY=~/pkey.key \
GLOO_DEVICE_TRANSPORT_TCP_TLS_CERT=~/cert.pem \
GLOO_DEVICE_TRANSPORT_TCP_TLS_CA_FILE=~/ca.pem \
./build/bin/ProcessGroupGlooAsyncTest
```




[ghstack-poisoned]

pbelevich committed:
Update on "Add Gloo OpenSSL-over-TCP transport"

This PR adds [Gloo OpenSSL-over-TCP transport](https://github.com/facebookincubator/gloo/pull/260) to PyTorch. This functionality works only on Linux and [gh tool](https://github.com/cli/cli#installation) and [openssl tool](https://www.openssl.org/) are required for manual testing.

0. Generate `pkey.key`, `cert.pem` and `ca.pem` in the `$HOME` dir:
```
openssl genrsa -out ca.key 2048
openssl req -subj "/C=US/ST=New York/L=New York/O=Gloo Certificate Authority" -new -x509 -days 7300 -key ca.key -sha256 -extensions v3_ca -out ca.pem

openssl genrsa -out pkey.key 2048
openssl req -subj "/C=US/ST=California/L=San Francisco/O=Gloo Testing Company" -sha256 -new -key pkey.key -out csr.csr

openssl x509 -sha256 -req -in csr.csr -CA ca.pem -CAkey ca.key -CAcreateserial -out cert.pem -days 7300

openssl verify -CAfile ca.pem cert.pem
```
1. Checkout this PR (`gh pr checkout 43482`)
2. (skip if https://github.com/facebookincubator/gloo/pull/260 is already merged) Goto gloo subdir: `cd third_party/gloo`
3. (skip if https://github.com/facebookincubator/gloo/pull/260 is already merged) Checkout gloo's PR: `gh pr checkout 260` and go back `cd ../..`
4. Build PyTorch with `USE_TCP_OPENSSL=ON python setup.py develop`
5. Run ProcessGroupGlooTest:
```
GLOO_DEVICE_TRANSPORT=TCP_TLS \
GLOO_DEVICE_TRANSPORT_TCP_TLS_PKEY=~/pkey.key \
GLOO_DEVICE_TRANSPORT_TCP_TLS_CERT=~/cert.pem \
GLOO_DEVICE_TRANSPORT_TCP_TLS_CA_FILE=~/ca.pem \
./build/bin/ProcessGroupGlooTest
```
6. Run ProcessGroupGlooAsyncTest:
```
GLOO_DEVICE_TRANSPORT=TCP_TLS \
GLOO_DEVICE_TRANSPORT_TCP_TLS_PKEY=~/pkey.key \
GLOO_DEVICE_TRANSPORT_TCP_TLS_CERT=~/cert.pem \
GLOO_DEVICE_TRANSPORT_TCP_TLS_CA_FILE=~/ca.pem \
./build/bin/ProcessGroupGlooAsyncTest
```




[ghstack-poisoned]

pbelevich committed:
Update on "Add Gloo OpenSSL-over-TCP transport"

This PR adds [Gloo OpenSSL-over-TCP transport](https://github.com/facebookincubator/gloo/pull/260) to PyTorch. This functionality works only on Linux and [gh tool](https://github.com/cli/cli#installation) and [openssl tool](https://www.openssl.org/) are required for manual testing.

0. Generate `pkey.key`, `cert.pem` and `ca.pem` in the `$HOME` dir:
```
openssl genrsa -out ca.key 2048
openssl req -subj "/C=US/ST=New York/L=New York/O=Gloo Certificate Authority" -new -x509 -days 7300 -key ca.key -sha256 -extensions v3_ca -out ca.pem

openssl genrsa -out pkey.key 2048
openssl req -subj "/C=US/ST=California/L=San Francisco/O=Gloo Testing Company" -sha256 -new -key pkey.key -out csr.csr

openssl x509 -sha256 -req -in csr.csr -CA ca.pem -CAkey ca.key -CAcreateserial -out cert.pem -days 7300

openssl verify -CAfile ca.pem cert.pem
```
1. Checkout this PR (`gh pr checkout 43482`)
2. (skip if https://github.com/facebookincubator/gloo/pull/260 is already merged) Goto gloo subdir: `cd third_party/gloo`
3. (skip if https://github.com/facebookincubator/gloo/pull/260 is already merged) Checkout gloo's PR: `gh pr checkout 260` and go back `cd ../..`
4. Build PyTorch with `USE_TCP_OPENSSL=ON python setup.py develop`
5. Run ProcessGroupGlooTest:
```
GLOO_DEVICE_TRANSPORT=TCP_TLS \
GLOO_DEVICE_TRANSPORT_TCP_TLS_PKEY=~/pkey.key \
GLOO_DEVICE_TRANSPORT_TCP_TLS_CERT=~/cert.pem \
GLOO_DEVICE_TRANSPORT_TCP_TLS_CA_FILE=~/ca.pem \
./build/bin/ProcessGroupGlooTest
```
6. Run ProcessGroupGlooAsyncTest:
```
GLOO_DEVICE_TRANSPORT=TCP_TLS \
GLOO_DEVICE_TRANSPORT_TCP_TLS_PKEY=~/pkey.key \
GLOO_DEVICE_TRANSPORT_TCP_TLS_CERT=~/cert.pem \
GLOO_DEVICE_TRANSPORT_TCP_TLS_CA_FILE=~/ca.pem \
./build/bin/ProcessGroupGlooAsyncTest
```




[ghstack-poisoned]

pbelevich commented:
@malfet @seemethere I will need your help with steps 4 and 5 in the PR description

pbelevich committed:
Update on "Add Gloo OpenSSL-over-TCP transport"

This PR adds [Gloo OpenSSL-over-TCP transport](https://github.com/facebookincubator/gloo/pull/260) to PyTorch. 

- [x] Update `third_party/gloo` to [OpenSSL-over-TCP transport](01e2c2660cd43963ce1fe3e21220ac01f07d9a4b)
- [x] `find_package(OpenSSL REQUIRED)` and building Gloo with `USE_TCP_OPENSSL`
- [x] Update `torch/lib/c10d/GlooDeviceFactory.cpp` with new `TCP_TLS` key that will create Gloo TLS transport
- [ ] Make OpenSSL(libssl-dev) available on all CI linux environments, so CMake would be able to find it
- [ ] Add new CI test(not build) workflow that 
  - Gets fresh pytorch build
  - Generates `pkey.key`, `cert.pem` and `ca.pem`:

```
openssl genrsa -out ca.key 2048
openssl req -subj "/C=US/ST=New York/L=New York/O=Gloo Certificate Authority" -new -x509 -days 7300 -key ca.key -sha256 -extensions v3_ca -out ca.pem

openssl genrsa -out pkey.key 2048
openssl req -subj "/C=US/ST=California/L=San Francisco/O=Gloo Testing Company" -sha256 -new -key pkey.key -out csr.csr

openssl x509 -sha256 -req -in csr.csr -CA ca.pem -CAkey ca.key -CAcreateserial -out cert.pem -days 7300

openssl verify -CAfile ca.pem cert.pem
```
  - Runs `./build/bin/ProcessGroupGlooTest`, `./build/bin/ProcessGroupGlooAsyncTest` and `pytorch/test/distributed/test*.py` with the following environment variables:

```
GLOO_DEVICE_TRANSPORT=TCP_TLS \
GLOO_DEVICE_TRANSPORT_TCP_TLS_PKEY=~/pkey.key \
GLOO_DEVICE_TRANSPORT_TCP_TLS_CERT=~/cert.pem \
GLOO_DEVICE_TRANSPORT_TCP_TLS_CA_FILE=~/ca.pem \
```




[ghstack-poisoned]

pbelevich committed:
Update on "Add Gloo OpenSSL-over-TCP transport"

This PR adds [Gloo OpenSSL-over-TCP transport](https://github.com/facebookincubator/gloo/pull/260) to PyTorch. 

- [x] Update `third_party/gloo` to [OpenSSL-over-TCP transport](01e2c2660cd43963ce1fe3e21220ac01f07d9a4b)
- [x] `find_package(OpenSSL REQUIRED)` and building Gloo with `USE_TCP_OPENSSL`
- [x] Update `torch/lib/c10d/GlooDeviceFactory.cpp` with new `TCP_TLS` key that will create Gloo TLS transport
- [ ] Make OpenSSL(libssl-dev) available on all CI linux environments, so CMake would be able to find it
- [ ] Add new CI test(not build) workflow that 
  - Gets fresh pytorch build
  - Generates `pkey.key`, `cert.pem` and `ca.pem`:

```
openssl genrsa -out ca.key 2048
openssl req -subj "/C=US/ST=New York/L=New York/O=Gloo Certificate Authority" -new -x509 -days 7300 -key ca.key -sha256 -extensions v3_ca -out ca.pem

openssl genrsa -out pkey.key 2048
openssl req -subj "/C=US/ST=California/L=San Francisco/O=Gloo Testing Company" -sha256 -new -key pkey.key -out csr.csr

openssl x509 -sha256 -req -in csr.csr -CA ca.pem -CAkey ca.key -CAcreateserial -out cert.pem -days 7300

openssl verify -CAfile ca.pem cert.pem
```
  - Runs `./build/bin/ProcessGroupGlooTest`, `./build/bin/ProcessGroupGlooAsyncTest` and `pytorch/test/distributed/test*.py` with the following environment variables:

```
GLOO_DEVICE_TRANSPORT=TCP_TLS \
GLOO_DEVICE_TRANSPORT_TCP_TLS_PKEY=~/pkey.key \
GLOO_DEVICE_TRANSPORT_TCP_TLS_CERT=~/cert.pem \
GLOO_DEVICE_TRANSPORT_TCP_TLS_CA_FILE=~/ca.pem \
```




[ghstack-poisoned]

pbelevich committed:
Update on "Add Gloo OpenSSL-over-TCP transport"

This PR adds [Gloo OpenSSL-over-TCP transport](https://github.com/facebookincubator/gloo/pull/260) to PyTorch. 

- [x] Update `third_party/gloo` to [OpenSSL-over-TCP transport](01e2c2660cd43963ce1fe3e21220ac01f07d9a4b)
- [x] `find_package(OpenSSL REQUIRED)` and building Gloo with `USE_TCP_OPENSSL`
- [x] Update `torch/lib/c10d/GlooDeviceFactory.cpp` with new `TCP_TLS` key that will create Gloo TLS transport
- [ ] Make OpenSSL(libssl-dev) available on all CI linux environments, so CMake would be able to find it
- [ ] Add new CI test(not build) workflow that 
  - Gets fresh pytorch build
  - Generates `pkey.key`, `cert.pem` and `ca.pem`:

```
openssl genrsa -out ca.key 2048
openssl req -subj "/C=US/ST=New York/L=New York/O=Gloo Certificate Authority" -new -x509 -days 7300 -key ca.key -sha256 -extensions v3_ca -out ca.pem

openssl genrsa -out pkey.key 2048
openssl req -subj "/C=US/ST=California/L=San Francisco/O=Gloo Testing Company" -sha256 -new -key pkey.key -out csr.csr

openssl x509 -sha256 -req -in csr.csr -CA ca.pem -CAkey ca.key -CAcreateserial -out cert.pem -days 7300

openssl verify -CAfile ca.pem cert.pem
```
  - Runs `./build/bin/ProcessGroupGlooTest`, `./build/bin/ProcessGroupGlooAsyncTest` and `pytorch/test/distributed/test*.py` with the following environment variables:

```
GLOO_DEVICE_TRANSPORT=TCP_TLS \
GLOO_DEVICE_TRANSPORT_TCP_TLS_PKEY=~/pkey.key \
GLOO_DEVICE_TRANSPORT_TCP_TLS_CERT=~/cert.pem \
GLOO_DEVICE_TRANSPORT_TCP_TLS_CA_FILE=~/ca.pem \
```




[ghstack-poisoned]

pbelevich committed:
Update on "Add Gloo OpenSSL-over-TCP transport"

This PR adds [Gloo OpenSSL-over-TCP transport](https://github.com/facebookincubator/gloo/pull/260) to PyTorch. 

- [x] Update `third_party/gloo` to [OpenSSL-over-TCP transport](01e2c2660cd43963ce1fe3e21220ac01f07d9a4b)
- [x] `find_package(OpenSSL REQUIRED)` and building Gloo with `USE_TCP_OPENSSL`
- [x] Update `torch/lib/c10d/GlooDeviceFactory.cpp` with new `TCP_TLS` key that will create Gloo TLS transport
- [ ] Make OpenSSL(libssl-dev) available on all CI linux environments, so CMake would be able to find it
- [ ] Add new CI test(not build) workflow that 
  - Gets fresh pytorch build
  - Generates `pkey.key`, `cert.pem` and `ca.pem`:

```
openssl genrsa -out ca.key 2048
openssl req -subj "/C=US/ST=New York/L=New York/O=Gloo Certificate Authority" -new -x509 -days 7300 -key ca.key -sha256 -extensions v3_ca -out ca.pem

openssl genrsa -out pkey.key 2048
openssl req -subj "/C=US/ST=California/L=San Francisco/O=Gloo Testing Company" -sha256 -new -key pkey.key -out csr.csr

openssl x509 -sha256 -req -in csr.csr -CA ca.pem -CAkey ca.key -CAcreateserial -out cert.pem -days 7300

openssl verify -CAfile ca.pem cert.pem
```
  - Runs `./build/bin/ProcessGroupGlooTest`, `./build/bin/ProcessGroupGlooAsyncTest` and `pytorch/test/distributed/test*.py` with the following environment variables:

```
GLOO_DEVICE_TRANSPORT=TCP_TLS \
GLOO_DEVICE_TRANSPORT_TCP_TLS_PKEY=~/pkey.key \
GLOO_DEVICE_TRANSPORT_TCP_TLS_CERT=~/cert.pem \
GLOO_DEVICE_TRANSPORT_TCP_TLS_CA_FILE=~/ca.pem \
```




[ghstack-poisoned]

pbelevich committed:
Update on "Add Gloo OpenSSL-over-TCP transport"

This PR adds [Gloo OpenSSL-over-TCP transport](https://github.com/facebookincubator/gloo/pull/260) to PyTorch. 

- [x] Update `third_party/gloo` to [OpenSSL-over-TCP transport](01e2c2660cd43963ce1fe3e21220ac01f07d9a4b)
- [x] `find_package(OpenSSL REQUIRED)` and building Gloo with `USE_TCP_OPENSSL`
- [x] Update `torch/lib/c10d/GlooDeviceFactory.cpp` with new `TCP_TLS` key that will create Gloo TLS transport
- [ ] Make OpenSSL(libssl-dev) available on all CI linux environments, so CMake would be able to find it
- [ ] Add new CI test(not build) workflow that 
  - Gets fresh pytorch build
  - Generates `pkey.key`, `cert.pem` and `ca.pem`:

```
openssl genrsa -out ca.key 2048
openssl req -subj "/C=US/ST=New York/L=New York/O=Gloo Certificate Authority" -new -x509 -days 7300 -key ca.key -sha256 -extensions v3_ca -out ca.pem

openssl genrsa -out pkey.key 2048
openssl req -subj "/C=US/ST=California/L=San Francisco/O=Gloo Testing Company" -sha256 -new -key pkey.key -out csr.csr

openssl x509 -sha256 -req -in csr.csr -CA ca.pem -CAkey ca.key -CAcreateserial -out cert.pem -days 7300

openssl verify -CAfile ca.pem cert.pem
```
  - Runs `./build/bin/ProcessGroupGlooTest`, `./build/bin/ProcessGroupGlooAsyncTest` and `pytorch/test/distributed/test*.py` with the following environment variables:

```
GLOO_DEVICE_TRANSPORT=TCP_TLS \
GLOO_DEVICE_TRANSPORT_TCP_TLS_PKEY=~/pkey.key \
GLOO_DEVICE_TRANSPORT_TCP_TLS_CERT=~/cert.pem \
GLOO_DEVICE_TRANSPORT_TCP_TLS_CA_FILE=~/ca.pem \
```




[ghstack-poisoned]
